doc/reference/trd-uart.md

```diff
@@ -0,0 +1,527 @@
+using 8-bit data words, each data word is a byte. If the UART is using
+smaller data words, it MUST ignore the high order bits of the data
+values. For example, if the UART is using 6-bit data words and is told
+to transmit `0xff`, it will transmit `0x3f`, ignoring the first two
+bits. If a client needs to transmit data words larger than 8 bits, it
+should use `transmit_word`, as `transmit_buffer` is a buffer of 8-bit
+bytes and cannot store 9-bit values.
+
+There can be a single transmit operation ongoing at any
+time. Successfully calling either `transmit_buffer` or
+`transmit_word` causes the UART to become busy until it issues the
+callback corresponding to the outstanding operation.
+
+3.1 `transmit_buffer` and `transmitted_buffer`
+===============================
+
+`Transmit::transmit_buffer` sends a buffer of data. The result
+returned by `transmit_buffer` indicates whether there will be a
+callback in the future. If `transmit_buffer` returns `Ok(())`,
+implementation MUST call the `TransmitClient::transmitted_buffer`
+callback in the future when the transmission completes or fails. If
+`transmit_buffer` returns `Err` it MUST NOT issue a callback in the
+future in response to this call. If the error is `BUSY`, this is
+because there is an outstanding call to `transmit_buffer` or
+`transmit_word`: the implementation will continue to handle
+the original call and issue the originally scheduled callback
+(as if the call that `Err`'d with `BUSY` never happened). However,
+ it does not issue a callback for the call to `transmit_buffer`
+  that returned `Err`.
+
+The valid error codes for `transmit_buffer` are:
+  - `OFF`: the underlying hardware is not available, perhaps because it has
+    not been initialized or has been initialized into a different mode
+    (e.g., a USART has been configured to be a SPI).
+  - `BUSY`: the UART is already transmitting and has not made a transmission
+    complete callback yet.
+  - `SIZE`: `tx_len` is larger than the passed slice.
+  - `FAIL`: some other failure.
+
+Calling `transmit_buffer` while there is an outstanding
+`transmit_buffer` or `transmit_word` operation MUST return `Err(BUSY)`.
+
+The `TransmitClient::transmitted_buffer` callback indicates completion
+of a buffer transmission.  The `Result` indicates whether the buffer
+was successfully transmitted.  The `tx_len` argument specifies how
+many data words (defined by `Configure`) were transmitted. If the
+`rval` of `transmitted_buffer` is `Ok(())`, `tx_len` MUST be equal to
+the size of the transmission started by `transmit_buffer`, defined
+above.  A call to `transmit_word` or `transmit_buffer` made within
+this callback MUST NOT return `Err(BUSY)` unless it is because this is
+not the first call to one of these methods in the callback.  When this
+callback is made, the UART MUST be ready to receive another call. The
+valid `ErrorCode` values for `transmitted_buffer` are all of those
+returned by `transmit_buffer` plus:
```

Comment:
Do all of those errors make sense for the already-started callback case?
 - `OFF`: I can see this if some higher priority task pre-empted your HW access
 - `BUSY`: Same priority thought?
 - `SIZE`: I don't think this error should be possible given the narrow definition above (the slice length is known at call-time)
 - `FAIL`: Sure.

So maybe "except `SIZE`"?